src/qutip_qoc/_rl.py

```diff
@@ -0,0 +1,313 @@
+        self._result.end_local_time = time.strftime("%Y-%m-%d %H:%M:%S", self._result.end_local_time)           # Convert to a string
+        self._result._guess_controls = []
+        self._result._optimized_H = [self._H]
+        self._result.guess_params = []
+        return self._result
+
+    def train(self):
+        """
+        Train the RL agent on the defined quantum control problem using the specified 
+        reinforcement learning algorithm. Checks environment compatibility with Gym API.
+        """
+        # Check if the environment follows Gym API
+        check_env(self, warn=True)
+
+        # Create the model
+        model = PPO('MlpPolicy', self, verbose=1)       # verbose = 1 to display training progress and statistics in the terminal
+        
+        stop_callback = EarlyStopTraining(verbose=1)
+        
+        # Train the model
+        model.learn(total_timesteps = self.total_timesteps, callback=stop_callback)
+
+class EarlyStopTraining(BaseCallback):
+    """
+    A callback to stop training based on specific conditions (steps, infidelity, max iterations)
+    """
+    def __init__(self, verbose: int = 0):
+        super(EarlyStopTraining, self).__init__(verbose)
+        self.stop_train = False
+
+    def _on_step(self) -> bool:
+        """
+        This method is required by the BaseCallback class. We use it only to stop the training.
+        """
+        # Check if we need to stop training
+        if self.stop_train:
+            return False  # Stop training
+        return True  # Continue training
+
+    def _on_rollout_start(self) -> None:
+        """
+        This method is called before the rollout starts (before collecting new samples).
+        Checks:
+        - If all of the last 100 episodes have infidelity below the target and use the same number of steps, stop training.
+        - Stop training if the maximum number of episodes is reached.
+        """
+        env = self.training_env.envs[0].unwrapped
+        max_episodes = env.max_episodes
+        fid_err_targ = env._fid_err_targ
+
+        if len(env.episode_info) >= 100:
+            last_100_episodes = env.episode_info[-100:]
+
+            min_steps = min(info['steps_used'] for info in last_100_episodes)
+            steps_condition = all(ep['steps_used'] == min_steps for ep in last_100_episodes)
+            infid_condition = all(ep['final_infidelity'] <= fid_err_targ for ep in last_100_episodes)
+
+            if steps_condition and infid_condition:
+                env._result.message = "Training finished. No episode in the last 100 used fewer steps and infidelity was below target infid."
+                #print(f"Stopping training as no episode in the last 100 used fewer steps and infidelity was below target infid.")
+                self.stop_train = True  # Stop training
+                #print([ep['steps_used'] for ep in last_100_episodes])
+                #print([ep['final_infidelity'] for ep in last_100_episodes])
+
```

Comment:
Please also add a abort condition to stop training once the fidelity target has been reached.
You can introduce an `alg_kwarg` to disable this option and have the RL agent look for shorter pulses.